File: Algorithms/sorting/quicksort.py
"""
快速排序 - Python 版
=====================
文件：quicksort.py
作者：你的名字
日期：2026

Python 版快速排序，与 C 版对比学习。
特点：Python 代码更简洁，但性能不及C（解释型语言）。
AI 应用：排序是数据预处理的基础操作。
"""

import time
import random
from typing import List

# ===== 版本1：最简洁版（利用 Python 列表推导式）=====

def quicksort_simple(arr: List[int]) -> List[int]:
    """
    最 Python 风格的快排（简洁但创建了新数组，空间 O(n log n)）
    这版本虽然清晰，但不是原地排序
    """
    if len(arr) <= 1:
        return arr

    pivot = arr[len(arr) // 2]      # 取中间元素作为 pivot
    left  = [x for x in arr if x < pivot]
    mid   = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]

    return quicksort_simple(left) + mid + quicksort_simple(right)


# ===== 版本2：原地快排（节省内存）=====

def _partition(arr: List[int], low: int, high: int) -> int:
    """Lomuto 分区方案"""
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]   # Python 交换无需临时变量！

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def quicksort_inplace(arr: List[int], low: int = 0, high: int = None) -> None:
    """原地快速排序（修改原数组）"""
    if high is None:
        high = len(arr) - 1

    if low < high:
        pi = _partition(arr, low, high)
        quicksort_inplace(arr, low, pi - 1)
        quicksort_inplace(arr, pi + 1, high)


# ===== 版本3：迭代版（避免递归栈溢出）=====

def quicksort_iterative(arr: List[int]) -> None:
    """
    用显式栈模拟递归，避免深度递归导致 Python 栈溢出
    Python 默认递归深度限制约 1000
    """
    if len(arr) <= 1:
        return

    stack = [(0, len(arr) - 1)]

    while stack:
        low, high = stack.pop()
        if low < high:
            pi = _partition(arr, low, high)
            stack.append((low, pi - 1))
            stack.append((pi + 1, high))


# ===== 对比 Python 内置排序 =====

def benchmark():
    """对比不同排序方法的速度"""
    n = 50000
    data = [random.randint(0, 10000) for _ in range(n)]

    print(f"测试规模: {n} 个元素\n")

    # 简洁版
    arr1 = data.copy()
    t0 = time.time()
    result1 = quicksort_simple(arr1)
    t1 = time.time() - t0
    print(f"简洁版快排:     {t1*1000:.1f}ms")

    # 原地版
    arr2 = data.copy()
    t0 = time.time()
    quicksort_inplace(arr2)
    t2 = time.time() - t0
    print(f"原地版快排:     {t2*1000:.1f}ms")

    # 迭代版
    arr3 = data.copy()
    t0 = time.time()
    quicksort_iterative(arr3)
    t3 = time.time() - t0
    print(f"迭代版快排:     {t3*1000:.1f}ms")

    # Python 内置（Timsort，混合归并+插入）
    arr4 = data.copy()
    t0 = time.time()
    arr4.sort()
    t4 = time.time() - t0
    print(f"Python sort():  {t4*1000:.1f}ms  ← 这是C实现，最快")

    # 验证正确性
    assert result1 == arr4, "简洁版排序错误！"
    assert arr2    == arr4, "原地版排序错误！"
    assert arr3    == arr4, "迭代版排序错误！"
    print("\n✅ 所有版本结果正确")


if __name__ == "__main__":
    # 小规模演示
    arr = [64, 34, 25, 12, 22, 11, 90, 7]
    print(f"原始: {arr}")
    sorted_arr = quicksort_simple(arr.copy())
    print(f"快排: {sorted_arr}")

    # 性能测试
    print("\n===== 性能对比 =====")
    benchmark()

"""
TODO 练习：

TODO-01: 实现归并排序（Python版）
    merge_sort(arr) 分治思想，结果正确且稳定

TODO-02: 用 heapq 实现堆排序
    import heapq; heapq.heapify(arr); ...

TODO-03: 实现基数排序（Radix Sort）
    先按个位排，再按十位排，... 最后得到有序数组

TODO-04: 可视化排序过程（用 matplotlib 动画！）
    每次交换后更新柱状图，视觉化理解排序

TODO-05: 实现归并排序计算逆序对数量
    对 [3, 1, 2] 而言，逆序对是 (3,1), (3,2)，共2个
"""

File: Algorithms/sorting/quicksort.c
/**
 * @file    quicksort.c
 * @brief   快速排序（Quick Sort）实现 - 分治法经典案例
 * @author  你的名字
 * @date    2026
 *
 * 快速排序：
 *   - 平均时间复杂度：O(n log n)
 *   - 最坏时间复杂度：O(n²)（已排序数组）
 *   - 空间复杂度：O(log n)（递归栈）
 *   - 不稳定排序
 *
 * 思想：选一个"基准"（pivot），把数组分成"比pivot小"和"比pivot大"两部分，
 *       然后递归地对两部分继续排序。
 *
 * 编译：gcc -Wall -std=c99 -o quicksort quicksort.c
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

/* ===== 工具函数 ===== */
void swap(int *a, int *b)
{
    int tmp = *a;
    *a = *b;
    *b = tmp;
}

void print_array(const int *arr, int n, const char *label)
{
    printf("%s: [", label);
    for (int i = 0; i < n; i++) {
        printf("%d%s", arr[i], (i < n - 1) ? ", " : "");
    }
    printf("]\n");
}

/* ===== 版本1：基础快速排序（Lomuto 分区）===== */

/**
 * @brief Lomuto 分区：选最后一个元素作为 pivot
 * @return 分区后 pivot 的最终位置
 */
int partition_lomuto(int *arr, int low, int high)
{
    int pivot = arr[high];    /* 选最后一个元素作为基准 */
    int i = low - 1;          /* i 指向"小于pivot区域"的边界 */

    for (int j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            swap(&arr[i], &arr[j]);   /* 小于等于pivot的移到左边 */
        }
    }
    swap(&arr[i + 1], &arr[high]);    /* pivot 放到最终位置 */
    return i + 1;
}

void quick_sort(int *arr, int low, int high)
{
    if (low < high) {
        int pi = partition_lomuto(arr, low, high);  /* 分区 */
        quick_sort(arr, low, pi - 1);               /* 递归左半部分 */
        quick_sort(arr, pi + 1, high);              /* 递归右半部分 */
    }
}

/* ===== 版本2：三路快排（处理大量重复元素效率更高）===== */

/**
 * @brief 三路分区：将数组分为"小于"、"等于"、"大于" pivot 三部分
 * @note  当有大量重复元素时，性能远好于普通快排
 */
void quick_sort_3way(int *arr, int low, int high)
{
    if (low >= high) return;

    int pivot = arr[low];
    int lt = low;         /* arr[low..lt-1]  < pivot */
    int gt = high;        /* arr[gt+1..high] > pivot */
    int i  = low + 1;     /* arr[lt..i-1]   == pivot */

    while (i <= gt) {
        if (arr[i] < pivot) {
            swap(&arr[lt], &arr[i]);
            lt++;
            i++;
        } else if (arr[i] > pivot) {
            swap(&arr[i], &arr[gt]);
            gt--;
        } else {
            i++;
        }
    }

    quick_sort_3way(arr, low, lt - 1);
    quick_sort_3way(arr, gt + 1, high);
}

/* ===== 版本3：随机化快排（避免最坏情况）===== */

void quick_sort_random(int *arr, int low, int high)
{
    if (low < high) {
        /* 随机选 pivot，与最后一个元素交换 */
        int rand_idx = low + rand() % (high - low + 1);
        swap(&arr[rand_idx], &arr[high]);

        int pi = partition_lomuto(arr, low, high);
        quick_sort_random(arr, low, pi - 1);
        quick_sort_random(arr, pi + 1, high);
    }
}

/* 用于 stdlib qsort 的比较函数（必须在文件作用域定义）*/
static int cmp_int(const void *a, const void *b)
{
    return *(const int *)a - *(const int *)b;
}

/* ===== 性能测试 ===== */
void benchmark_sort(int n)
{
    printf("\n===== 性能测试（n=%d）=====\n", n);

    int *arr1 = (int *)malloc(n * sizeof(int));
    int *arr2 = (int *)malloc(n * sizeof(int));

    /* 生成随机数据 */
    srand(42);
    for (int i = 0; i < n; i++) {
        arr1[i] = rand() % 10000;
    }
    memcpy(arr2, arr1, n * sizeof(int));

    /* 测试 quick_sort */
    clock_t t0 = clock();
    quick_sort(arr1, 0, n - 1);
    double t_qsort = (double)(clock() - t0) / CLOCKS_PER_SEC * 1000;

    /* 测试 stdlib qsort */
    t0 = clock();
    qsort(arr2, n, sizeof(int), cmp_int);
    double t_stdlib = (double)(clock() - t0) / CLOCKS_PER_SEC * 1000;

    printf("自实现快排: %.2fms\n", t_qsort);
    printf("stdlib qsort: %.2fms\n", t_stdlib);

    /* 验证正确性 */
    int correct = 1;
    for (int i = 0; i < n; i++) {
        if (arr1[i] != arr2[i]) { correct = 0; break; }
    }
    printf("排序结果: %s\n", correct ? "✓ 正确" : "✗ 错误！");

    free(arr1);
    free(arr2);
}

int main(void)
{
    srand((unsigned)time(NULL));

    /* 基础演示 */
    int arr[] = {64, 34, 25, 12, 22, 11, 90, 7};
    int n = sizeof(arr) / sizeof(arr[0]);

    print_array(arr, n, "排序前");
    quick_sort(arr, 0, n - 1);
    print_array(arr, n, "快排后");

    /* 三路快排（含重复元素）*/
    int arr2[] = {3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5};
    int n2 = sizeof(arr2) / sizeof(arr2[0]);
    print_array(arr2, n2, "\n三路快排前");
    quick_sort_3way(arr2, 0, n2 - 1);
    print_array(arr2, n2, "三路快排后");

    /* 性能测试 */
    benchmark_sort(100000);

    return 0;
}

/*
 * TODO 练习：
 *
 * TODO-01: 实现归并排序（Merge Sort）
 *   思想：分成两半，递归排序，然后合并两个有序数组
 *   时间：O(n log n)，稳定排序，但需要 O(n) 额外空间
 *
 * TODO-02: 实现堆排序（Heap Sort）
 *   先建最大堆，再反复把堆顶（最大值）与末尾交换
 *
 * TODO-03: 实现计数排序（Counting Sort）
 *   适用于整数范围不大的场景，时间 O(n+k)
 *
 * TODO-04: 对比各种排序的速度（绘制折线图，用Python+matplotlib）
 *
 * TODO-05: 快排优化：当子数组长度 < 10 时改用插入排序（工程实践）
 */
